test/jdk/java/lang/String/concat/HiddenClassUnloading.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.lang.StringBuilder;
+
+import java.lang.invoke.*;
+import java.lang.management.ManagementFactory;
+
+/**
+ * @test
+ * @summary Test whether the hidden class unloading of StringConcatFactory works
+ *
+ * @run main/othervm -Xmx8M -Xms8M -Xverify:all HiddenClassUnloading
+ * @run main/othervm -Xmx8M -Xms8M -Xverify:all -XX:-CompactStrings HiddenClassUnloading
+ */
+public class HiddenClassUnloading {
+    public static void main(String[] args) throws Throwable {
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+        Class<?>[] types = new Class[] {
+                int.class, long.class, double.class, float.class, char.class, boolean.class, String.class,
+        };
+        Object[] values = new Object[] {
+                1, 1L, 1D, 1F, 'C', true, "A",
+        };
+
+        for (int i = 0; i < 10_000; i++) {
+            int radix = types.length;
+            String str = Integer.toString(i, radix);
+            int length = str.length();
+            String recipe = "\1".repeat(length);
+            Class<?>[] ptypes = new Class[length];
+            Object[] pvalues = new Object[length];
+            for (int j = 0; j < length; j++) {
+                int index = Integer.parseInt(str.substring(j, j + 1), radix);
+                ptypes[j] = types[index];
+                pvalues[j] = values[index];
+            }
+            MethodType concatType = MethodType.methodType(String.class, ptypes);
+            CallSite callSite = StringConcatFactory.makeConcatWithConstants(
+                    lookup,
+                    "concat",
+                    concatType,
+                    recipe,
+                    new Object[0]
+            );
+            MethodHandle mh = callSite.dynamicInvoker();
```

Comment:
Actually invoking the concats seem unnecessary for this test; even with the rest of this method removed many thousands of classes is unloaded. We also seem to do pretty well with fewer iterations in the outer loop.

Reply:
Yes, this is not necessary, it is there to check correctness, but it is not what this test is about, so I have removed it.